File: packages/typescript/src/features/autonomy/action.ts
/**
 * Autonomy Actions for elizaOS
 *
 * Actions that enable autonomous agent communication.
 */

import { v4 as uuidv4 } from "uuid";
import {
	findKeywordTermMatch,
	getValidationKeywordTerms,
} from "../../i18n/validation-keywords.ts";
import type {
	Action,
	ActionResult,
	HandlerCallback,
	HandlerOptions,
	IAgentRuntime,
	Memory,
	State,
	UUID,
} from "../../types";
import { stringToUuid } from "../../utils";
import { AUTONOMY_SERVICE_TYPE, type AutonomyService } from "./service";

const ADMIN_KEYWORDS = getValidationKeywordTerms("action.sendToAdmin.request", {
	includeAllLocales: true,
});

/**
 * Send to Admin Action
 *
 * Allows agent to send messages to admin from autonomous context.
 * Only available in autonomous room to prevent misuse.
 */
export const sendToAdminAction: Action = {
	name: "SEND_TO_ADMIN",
	description:
		"Send a message directly to the admin user from autonomous context",

	examples: [
		[
			{
				name: "Agent",
				content: {
					text: "I need to update the admin about my progress on the task.",
					action: "SEND_TO_ADMIN",
				},
			},
			{
				name: "Agent",
				content: {
					text: "Message sent to admin successfully.",
				},
			},
		],
		[
			{
				name: "Agent",
				content: {
					text: "I should let the admin know I completed the analysis.",
					action: "SEND_TO_ADMIN",
				},
			},
			{
				name: "Agent",
				content: {
					text: "Admin has been notified of the analysis completion.",
				},
			},
		],
	],

	validate: async (
		runtime: IAgentRuntime,
		message: Memory,
	): Promise<boolean> => {
		// Only allow this action in autonomous context
		const autonomyService = runtime.getService<AutonomyService>(
			AUTONOMY_SERVICE_TYPE,
		);
		if (!autonomyService) {
			return false;
		}

		const autonomousRoomId = autonomyService.getAutonomousRoomId?.();
		if (!autonomousRoomId || message.roomId !== autonomousRoomId) {
			return false;
		}

		// Check if admin is configured
		const adminUserId = runtime.getSetting("ADMIN_USER_ID");
		if (typeof adminUserId !== "string" || adminUserId.length === 0) {
			return false;
		}

		// Check if message contains intention to communicate with admin
		const text = message.content.text ?? "";
		if (text.length === 0) {
			return false;
		}
		return findKeywordTermMatch(text, ADMIN_KEYWORDS) !== undefined;
	},

	handler: async (
		runtime: IAgentRuntime,
		message: Memory,
		_state?: State,
		_options?: HandlerOptions,
		callback?: HandlerCallback,
	): Promise<ActionResult> => {
		// Double-check we're in autonomous context
		const autonomyService = runtime.getService<AutonomyService>(
			AUTONOMY_SERVICE_TYPE,
		);
		if (!autonomyService) {
			return {
				success: false,
				text: "Autonomy service not available",
				data: { error: "Service unavailable" },
			};
		}

		const autonomousRoomId = autonomyService.getAutonomousRoomId?.();
		if (!autonomousRoomId || message.roomId !== autonomousRoomId) {
			return {
				success: false,
				text: "Send to admin only available in autonomous context",
				data: { error: "Invalid context" },
			};
		}

		// Get admin user ID
		const adminUserId = runtime.getSetting("ADMIN_USER_ID");
		if (typeof adminUserId !== "string" || adminUserId.length === 0) {
			return {
				success: false,
				text: "No admin user configured. Set ADMIN_USER_ID in settings.",
				data: { error: "No admin configured" },
			};
		}

		// Find target room
		const adminMessages = await runtime.getMemories({
			roomId: runtime.agentId,
			limit: 10,
			tableName: "memories",
		});

		let targetRoomId: UUID;
		if (adminMessages && adminMessages.length > 0) {
			const lastMessage = adminMessages[adminMessages.length - 1];
			targetRoomId = lastMessage.roomId ?? runtime.agentId;
		} else {
			targetRoomId = runtime.agentId;
		}

		// Extract message content
		const autonomousThought = message.content.text || "";

		// Generate message to admin
		let messageToAdmin: string;
		if (
			autonomousThought.includes("completed") ||
			autonomousThought.includes("finished")
		) {
			messageToAdmin = `I've completed a task and wanted to update you. My thoughts: ${autonomousThought}`;
		} else if (
			autonomousThought.includes("problem") ||
			autonomousThought.includes("issue") ||
			autonomousThought.includes("error")
		) {
			messageToAdmin = `I encountered something that might need your attention: ${autonomousThought}`;
		} else if (
			autonomousThought.includes("question") ||
			autonomousThought.includes("unsure")
		) {
			messageToAdmin = `I have a question and would appreciate your guidance: ${autonomousThought}`;
		} else {
			messageToAdmin = `Autonomous update: ${autonomousThought}`;
		}

		// Create and store message
		const now = Date.now();
		const adminMessage: Memory = {
			id: stringToUuid(uuidv4()),
			entityId: runtime.agentId,
			roomId: targetRoomId,
			content: {
				text: messageToAdmin,
				source: "autonomy-to-admin",
				metadata: {
					type: "autonomous-to-admin-message",
					originalThought: autonomousThought,
					timestamp: now,
				},
			},
			createdAt: now,
		};

		await runtime.createMemory(adminMessage, "memories");

		const successMessage = `Message sent to admin in room ${targetRoomId.slice(0, 8)}...`;

		if (callback) {
			await callback({
				text: successMessage,
				data: {
					adminUserId,
					targetRoomId,
					messageContent: messageToAdmin,
				},
			});
		}

		return {
			success: true,
			text: successMessage,
			data: {
				adminUserId,
				targetRoomId,
				messageContent: messageToAdmin,
				sent: true,
			},
		};
	},
};

File: packages/rust/src/features/autonomy/action.rs
//! Autonomy actions.

use std::collections::HashMap;
use std::sync::{Arc, Weak};

use anyhow::Result;
use serde_json::Value;

use crate::autonomy::service::AutonomyService;
use crate::runtime::AgentRuntime;
use crate::types::components::{ActionDefinition, ActionHandler, ActionResult, HandlerOptions};
use crate::types::memory::Memory;
use crate::types::settings::SettingValue;
use crate::types::state::State;

/// SEND_TO_ADMIN action (restricted to autonomous room).
pub struct SendToAdminAction {
    runtime: Weak<AgentRuntime>,
    service: Arc<AutonomyService>,
}

impl SendToAdminAction {
    /// Create a new send-to-admin action handler.
    pub fn new(runtime: Weak<AgentRuntime>, service: Arc<AutonomyService>) -> Self {
        Self { runtime, service }
    }
}

#[async_trait::async_trait]
impl ActionHandler for SendToAdminAction {
    fn definition(&self) -> ActionDefinition {
        ActionDefinition {
            name: "SEND_TO_ADMIN".to_string(),
            description: "Send a message directly to the admin user from autonomous context"
                .to_string(),
            similes: None,
            examples: None,
            priority: None,
            tags: None,
            parameters: None,
        }
    }

    async fn validate(&self, message: &Memory, _state: Option<&State>) -> bool {
        // Only allow in autonomous room, and only if admin configured.
        if message.room_id != self.service.autonomous_room_id() {
            return false;
        }
        let Some(rt) = self.runtime.upgrade() else {
            return false;
        };
        matches!(rt.get_setting("ADMIN_USER_ID").await, Some(SettingValue::String(s)) if !s.trim().is_empty())
    }

    async fn handle(
        &self,
        message: &Memory,
        _state: Option<&State>,
        _options: Option<&HandlerOptions>,
    ) -> Result<Option<ActionResult>> {
        let Some(rt) = self.runtime.upgrade() else {
            return Ok(Some(ActionResult::failure("Runtime unavailable")));
        };

        if message.room_id != self.service.autonomous_room_id() {
            return Ok(Some(ActionResult::failure(
                "Send to admin only available in autonomous context",
            )));
        }

        let admin_user_id = match rt.get_setting("ADMIN_USER_ID").await {
            Some(SettingValue::String(s)) if !s.trim().is_empty() => s,
            _ => {
                return Ok(Some(ActionResult::failure(
                    "No admin user configured. Set ADMIN_USER_ID in settings.",
                )));
            }
        };

        let autonomous_thought = message.content.text.clone().unwrap_or_default();
        let message_to_admin = format!("Autonomous update: {}", autonomous_thought);

        if let Some(adapter) = rt.get_adapter() {
            let mut content = crate::types::primitives::Content {
                text: Some(message_to_admin.clone()),
                source: Some("autonomy-to-admin".to_string()),
                ..Default::default()
            };
            content.extra.insert(
                "adminUserId".to_string(),
                Value::String(admin_user_id.clone()),
            );

            let mut mem = Memory::new(rt.agent_id.clone(), rt.agent_id.clone(), content);
            mem.agent_id = Some(rt.agent_id.clone());
            let _ = adapter.create_memory(&mem, "messages").await?;
        }

        let mut data: HashMap<String, Value> = HashMap::new();
        data.insert("adminUserId".to_string(), Value::String(admin_user_id));
        data.insert("sent".to_string(), Value::Bool(true));

        Ok(Some(ActionResult {
            success: true,
            text: Some("Message sent to admin".to_string()),
            values: None,
            data: Some(data),
            error: None,
        }))
    }
}
